modules/gdscript/gdscript_analyzer.cpp

```diff
@@ -403,30 +435,33 @@ Error GDScriptAnalyzer::resolve_inheritance(GDScriptParser::ClassNode *p_class,
 	class_type.native_type = result.native_type;
 	p_class->set_datatype(class_type);
 
-	if (p_recursive) {
-		for (int i = 0; i < p_class->members.size(); i++) {
-			if (p_class->members[i].type == GDScriptParser::ClassNode::Member::CLASS) {
-				Error err = resolve_inheritance(p_class->members[i].m_class, true);
-				if (err) {
-					return err;
-				}
-			}
-		}
-	}
+	parser->current_class = previous_class;
 
 	return OK;
 }
 
 GDScriptParser::DataType GDScriptAnalyzer::resolve_datatype(GDScriptParser::TypeNode *p_type) {
-	GDScriptParser::DataType result;
+	GDScriptParser::DataType bad_type;
+	bad_type.kind = GDScriptParser::DataType::VARIANT;
+	bad_type.type_source = GDScriptParser::DataType::INFERRED;
 
 	if (p_type == nullptr) {
-		result.kind = GDScriptParser::DataType::VARIANT;
-		return result;
+		return bad_type;
 	}
 
+	if (p_type->get_datatype().is_resolving()) {
+		push_error(R"(Could not resolve datatype: Cyclic reference.)", p_type);
+		return bad_type;
+	}
+
+	if (!p_type->get_datatype().has_no_type()) {
+		return p_type->get_datatype();
+	}
+
+	GDScriptParser::DataType result;
+	result.kind = GDScriptParser::DataType::RESOLVING;
+	p_type->set_datatype(result);
 	result.type_source = result.ANNOTATED_EXPLICIT;
```

Comment:
Minor issue with original code (not even yours), but isn't line 464 more correctly stated as the following?
`result.type_source = GDScriptParser::DataType::ANNOTATED_EXPLICIT;` rather than `result.ANNOTATED_EXPLICIT;`?

Reply:
yep you're right, i'll change that